File: server/src/content.ts
import { z } from "zod";
import { Prisma } from "@prisma/client";
import { prisma } from "./db.js";

/**
 * Case- and punctuation-insensitive form of a word, used as the course-wide
 * uniqueness key. "Hallo", "hallo" and "Hallo!" all normalise to the same
 * value, so none of them can be added twice under a different spelling.
 */
/** The original, file-based course. Courses made in the builder use their
 * own Course id, so words never clash across independent courses. */
export const LEGACY_COURSE_ID = "legacy";

export function normalizeWord(german: string): string {
  return german
    .toLowerCase()
    .replace(/[.,!?;:…"'«»„""()]/g, "")
    .replace(/\s+/g, " ")
    .trim();
}

/** Formal marks stripped only from the edges of a quiz answer/option — never
 * from the middle, where punctuation like a mid-phrase comma usually carries
 * real meaning. Mirrors the client's cleanQuizText (src/content/textUtils.ts)
 * exactly, since both must agree on what counts as a "hint" character. */
const QUIZ_EDGE_PUNCTUATION = /^[.!?;:…"'«»„""()/]+|[.!?;:…"'«»„""()/]+$/g;

/**
 * Cleans a string before it's stored as a quiz option/answer, so its
 * punctuation can never visually single out the correct choice among
 * distractors. Never applied to material text or a question's own prompt —
 * only to the answer-bearing fields validated in blockQuestionSchema
 * (courses.ts).
 */
export function cleanQuizText(value: string): string {
  return value.replace(QUIZ_EDGE_PUNCTUATION, "").replace(/\s+/g, " ").trim();
}

/** Human-readable lesson name for duplicate messages ("lesson2" -> "Урок 2"). */
export function lessonLabel(lessonId: string): string {
  const n = lessonId.match(/(\d+)$/);
  return n ? `Урок ${n[1]}` : lessonId;
}

/** Raised when a word would end up in two lessons of the same course. */
export class DuplicateWordError extends Error {}

export const QUESTION_SETS = ["minitest", "practice", "review"] as const;
export type QuestionSet = (typeof QUESTION_SETS)[number];

export const vocabularyItemSchema = z.object({
  german: z.string().trim().min(1, "Немецкое слово не может быть пустым"),
  translation: z.string().trim().min(1, "Перевод не может быть пустым"),
  pronunciation: z.string().trim().optional().nullable(),
  audioUrl: z.string().trim().optional().nullable(),
});

export const questionSchema = z
  .object({
    setName: z.enum(QUESTION_SETS),
    prompt: z.string().trim().min(1, "Текст вопроса не может быть пустым"),
    options: z.array(z.string().trim().min(1, "Вариант ответа не может быть пустым")).min(2, "Нужно минимум 2 варианта ответа"),
    correctAnswer: z.string().trim().min(1, "Нужно отметить правильный вариант"),
  })
  // Correctness is stored by value, so the marked answer must actually be one
  // of the options — otherwise the question would be unanswerable.
  .refine((q) => q.options.includes(q.correctAnswer), {
    message: "Правильный ответ должен быть одним из вариантов",
  })
  .refine((q) => new Set(q.options).size === q.options.length, {
    message: "Варианты ответа не должны повторяться",
  });

export const contentPayloadSchema = z.object({
  materialText: z.string().optional(),
  vocabulary: z.array(vocabularyItemSchema).optional(),
  questions: z.array(questionSchema).optional(),
});

export type ContentPayload = z.infer<typeof contentPayloadSchema>;

/**
 * A question authored in the admin panel (either course), tagged with its
 * kind. `kind` matches `Exercise["kind"]` in src/content/exercises.ts
 * exactly, so the client can map one straight onto the other with a plain
 * switch. Lives here (not in courses.ts) so both the legacy course and the
 * builder can import it without a circular dependency between the two.
 */
export type QuestionDTO =
  | { kind: "choice"; prompt: string; options: string[]; correctAnswer: string }
  | { kind: "truefalse"; prompt: string; correct: boolean }
  | { kind: "cloze"; prompt: string; options: string[]; correctAnswer: string }
  | { kind: "scramble"; prompt: string; options: string[]; correctAnswer: string }
  | { kind: "match"; prompt: string; pairs: { left: string; right: string }[] };

export function toQuestionDTO(q: {
  kind: string;
  prompt: string;
  options: string[];
  correctAnswer: string;
  data: unknown;
}): QuestionDTO {
  switch (q.kind) {
    case "truefalse":
      return { kind: "truefalse", prompt: q.prompt, correct: q.correctAnswer === "true" };
    case "cloze":
      return { kind: "cloze", prompt: q.prompt, options: q.options, correctAnswer: q.correctAnswer };
    case "scramble":
      return { kind: "scramble", prompt: q.prompt, options: q.options, correctAnswer: q.correctAnswer };
    case "match":
      return { kind: "match", prompt: q.prompt, pairs: (q.data as { left: string; right: string }[] | null) ?? [] };
    case "choice":
    default:
      return { kind: "choice", prompt: q.prompt, options: q.options, correctAnswer: q.correctAnswer };
  }
}

export interface LessonBlockDTO {
  id: string;
  stage: QuestionSet;
  title: string;
  position: number;
  questions: QuestionDTO[];
}

export interface LessonContentDTO {
  lessonId: string;
  /** null means "no admin edits yet" — the client keeps using the file. */
  materialText: string | null;
  /** null means "use the bundled file" — same override-if-present meaning as materialText. */
  videoUrl: string | null;
  audioUrl: string | null;
  vocabulary: { id: string; german: string; translation: string; pronunciation: string | null; audioUrl: string | null }[];
  questions: { setName: string; prompt: string; options: string[]; correctAnswer: string }[];
  /** Named question blocks (multiple mini-tests/practice/review sets), same shape the builder uses. */
  blocks: LessonBlockDTO[];
  /** False when nothing has ever been saved for this lesson. */
  hasOverrides: boolean;
  updatedAt: string | null;
  /** Saved LessonFlowCanvas layout, or null if never saved yet. */
  canvasLayout: unknown;
}

export async function getLessonContent(lessonId: string): Promise<LessonContentDTO> {
  const [content, vocabulary, questions, blocks] = await Promise.all([
    prisma.lessonContent.findUnique({ where: { lessonId } }),
    prisma.vocabularyItem.findMany({ where: { lessonId }, orderBy: { position: "asc" } }),
    prisma.lessonQuestion.findMany({ where: { lessonId }, orderBy: [{ setName: "asc" }, { position: "asc" }] }),
    prisma.lessonBlock.findMany({ where: { lessonId }, orderBy: [{ stage: "asc" }, { position: "asc" }] }),
  ]);

  return {
    lessonId,
    materialText: content?.materialText ?? null,
    videoUrl: content?.videoUrl ?? null,
    audioUrl: content?.audioUrl ?? null,
    vocabulary: vocabulary.map((v) => ({
      id: v.id,
      german: v.german,
      translation: v.translation,
      pronunciation: v.pronunciation,
      audioUrl: v.audioUrl,
    })),
    questions: questions.map((q) => ({
      setName: q.setName,
      prompt: q.prompt,
      options: q.options,
      correctAnswer: q.correctAnswer,
    })),
    blocks: blocks.map((b) => ({
      id: b.id,
      stage: b.stage as QuestionSet,
      title: b.title,
      position: b.position,
      questions: questions.filter((q) => q.blockId === b.id).map(toQuestionDTO),
    })),
    hasOverrides: Boolean(content) || vocabulary.length > 0 || questions.length > 0,
    updatedAt: content?.updatedAt.toISOString() ?? null,
    canvasLayout: content?.canvasLayout ?? null,
  };
}

/**
 * Saves whichever sections were supplied. Each section is replaced wholesale
 * inside one transaction, so a half-applied vocabulary list can never reach
 * learners. Sections that are omitted are left untouched.
 */
export async function saveLessonContent(
  lessonId: string,
  payload: ContentPayload,
  editorId: string,
): Promise<LessonContentDTO> {
  await prisma.$transaction(async (tx) => {
    if (payload.materialText !== undefined) {
      await tx.lessonContent.upsert({
        where: { lessonId },
        update: { materialText: payload.materialText, updatedById: editorId },
        create: { lessonId, materialText: payload.materialText, updatedById: editorId },
      });
    }

    if (payload.vocabulary !== undefined) {
      // A word may only belong to one lesson of the course. Reject the whole
      // save (rather than silently dropping a word) and name the lesson the
      // word already lives in, so the admin knows where to look.
      const keys = payload.vocabulary.map((item) => normalizeWord(item.german));
      const duplicateInPayload = keys.find((key, i) => keys.indexOf(key) !== i);
      if (duplicateInPayload) {
        const word = payload.vocabulary[keys.indexOf(duplicateInPayload)].german;
        throw new DuplicateWordError(`Слово «${word}» указано в этом уроке дважды`);
      }

      const clashes = await tx.vocabularyItem.findMany({
        where: { courseId: LEGACY_COURSE_ID, germanKey: { in: keys }, lessonId: { not: lessonId } },
        select: { german: true, lessonId: true },
      });
      if (clashes.length > 0) {
        const list = clashes
          .map((c) => `«${c.german}» — уже в уроке «${lessonLabel(c.lessonId)}»`)
          .join("; ");
        throw new DuplicateWordError(`Это слово уже используется в другом уроке: ${list}`);
      }

      // Recorded pronunciations are attached to the word, so they survive a
      // full rewrite of the lesson's list.
      const existing = await tx.vocabularyItem.findMany({
        where: { lessonId },
        select: { germanKey: true, audioUrl: true },
      });
      const audioByKey = new Map(existing.map((e) => [e.germanKey, e.audioUrl]));

      await tx.vocabularyItem.deleteMany({ where: { lessonId } });
      if (payload.vocabulary.length > 0) {
        await tx.vocabularyItem.createMany({
          data: payload.vocabulary.map((item, index) => {
            const germanKey = normalizeWord(item.german);
            return {
              lessonId,
              german: item.german,
              translation: item.translation,
              pronunciation: item.pronunciation?.trim() ? item.pronunciation.trim() : null,
              audioUrl: item.audioUrl?.trim() ? item.audioUrl.trim() : (audioByKey.get(germanKey) ?? null),
              position: index,
              germanKey,
            };
          }),
        });
      }
    }

    if (payload.questions !== undefined) {
      await tx.lessonQuestion.deleteMany({ where: { lessonId } });
      if (payload.questions.length > 0) {
        // Positions are per-set so each stage keeps its own ordering.
        const counters: Record<string, number> = {};
        await tx.lessonQuestion.createMany({
          data: payload.questions.map((q) => {
            const position = counters[q.setName] ?? 0;
            counters[q.setName] = position + 1;
            return {
              lessonId,
              setName: q.setName,
              prompt: q.prompt,
              options: q.options,
              correctAnswer: q.correctAnswer,
              position,
            };
          }),
        });
      }
    }
  });

  return getLessonContent(lessonId);
}

/**
 * Overrides (or clears, with url = null) the bundled video/audio file for a
 * legacy lesson. Upserts because a lesson may not have a LessonContent row
 * yet (e.g. its material has never been edited) — same pattern as
 * materialText/vocabulary above.
 */
export async function setLegacyLessonMedia(
  lessonId: string,
  kind: "video" | "audio",
  url: string | null,
): Promise<LessonContentDTO> {
  const field = kind === "video" ? "videoUrl" : "audioUrl";
  await prisma.lessonContent.upsert({
    where: { lessonId },
    update: { [field]: url },
    create: { lessonId, [field]: url },
  });
  return getLessonContent(lessonId);
}

/** Saves the LessonFlowCanvas layout for a legacy lesson — same upsert
 * pattern as setLegacyLessonMedia above, and just as presentation-only
 * (accepted as opaque JSON; see LessonContentDTO.canvasLayout). */
export async function saveLegacyCanvasLayout(lessonId: string, layout: unknown): Promise<LessonContentDTO> {
  await prisma.lessonContent.upsert({
    where: { lessonId },
    update: { canvasLayout: layout as Prisma.InputJsonValue },
    create: { lessonId, canvasLayout: layout as Prisma.InputJsonValue },
  });
  return getLessonContent(lessonId);
}

File: backend/app/services/content.py
import re

from sqlalchemy import and_, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.lesson_block import LessonBlock
from app.models.lesson_content import LessonContent
from app.models.lesson_question import LessonQuestion
from app.models.vocabulary_item import VocabularyItem
from app.schemas.content import ContentPayload
from app.services.material import filter_new_vocabulary, parse_material
from app.utils import to_iso_z, utcnow

# The original, file-based course. Courses made in the builder use their own
# Course id, so words never clash across independent courses.
LEGACY_COURSE_ID = "legacy"

_NORMALIZE_PUNCTUATION = re.compile(r"""[.,!?;:…"'«»„""()]""")
_WHITESPACE = re.compile(r"\s+")
_QUIZ_EDGE_PUNCTUATION = re.compile(r"""^[.!?;:…"'«»„""()/]+|[.!?;:…"'«»„""()/]+$""")
_TRAILING_DIGITS = re.compile(r"(\d+)$")


def normalize_word(german: str) -> str:
    """Case- and punctuation-insensitive form of a word, used as the
    course-wide uniqueness key. "Hallo", "hallo" and "Hallo!" all normalize
    to the same value, so none of them can be added twice under a different
    spelling."""
    value = german.lower()
    value = _NORMALIZE_PUNCTUATION.sub("", value)
    value = _WHITESPACE.sub(" ", value)
    return value.strip()


def clean_quiz_text(value: str) -> str:
    """Formal marks stripped only from the edges of a quiz answer/option —
    never from the middle. Mirrors the client's cleanQuizText exactly."""
    value = _QUIZ_EDGE_PUNCTUATION.sub("", value)
    value = _WHITESPACE.sub(" ", value)
    return value.strip()


def lesson_label(lesson_id: str) -> str:
    """Human-readable lesson name for duplicate messages ("lesson2" ->
    "Урок 2")."""
    m = _TRAILING_DIGITS.search(lesson_id)
    return f"Урок {m.group(1)}" if m else lesson_id


class DuplicateWordError(Exception):
    """Raised when a word would end up in two lessons of the same course."""


def to_question_dto(kind: str, prompt: str, options: list[str] | None, correct_answer: str, data) -> dict:
    if kind == "truefalse":
        return {"kind": "truefalse", "prompt": prompt, "correct": correct_answer == "true"}
    if kind == "cloze":
        return {"kind": "cloze", "prompt": prompt, "options": options or [], "correctAnswer": correct_answer}
    if kind == "scramble":
        return {"kind": "scramble", "prompt": prompt, "options": options or [], "correctAnswer": correct_answer}
    if kind == "match":
        return {"kind": "match", "prompt": prompt, "pairs": data or []}
    return {"kind": "choice", "prompt": prompt, "options": options or [], "correctAnswer": correct_answer}


async def _legacy_lesson_order(db: AsyncSession) -> list[str]:
    """Legacy lessons have no CourseLesson rows to order by position — the
    frontend's own file-discovery sorts by the trailing digit in the id
    (lesson1, lesson2, ...); replicated here from the set of lessonIds that
    actually have vocabulary, since that's the only place the backend can
    observe "which legacy lessons exist" without scanning files itself."""
    result = await db.execute(select(VocabularyItem.lessonId).where(VocabularyItem.courseId == LEGACY_COURSE_ID).distinct())
    ids = [r[0] for r in result.all()]

    def sort_key(lesson_id: str) -> int:
        m = _TRAILING_DIGITS.search(lesson_id)
        return int(m.group(1)) if m else 0

    return sorted(ids, key=sort_key)


async def _words_taught_before_legacy_lesson(db: AsyncSession, lesson_id: str) -> set[str]:
    order = await _legacy_lesson_order(db)
    if lesson_id not in order:
        return set()
    earlier_ids = order[: order.index(lesson_id)]
    if not earlier_ids:
        return set()
    result = await db.execute(
        select(VocabularyItem.germanKey).where(VocabularyItem.courseId == LEGACY_COURSE_ID, VocabularyItem.lessonId.in_(earlier_ids))
    )
    return {r[0] for r in result.all()}


def _title_from_blocks(blocks: list[dict]) -> str | None:
    for b in blocks:
        if b.get("type") == "title":
            return b.get("text")
    return None


async def list_legacy_lessons(db: AsyncSession) -> list[dict]:
    """All legacy lesson ids, in the same numeric order the old React
    client's file-discovery used to list them in. Needed by any Flutter
    screen that has to enumerate "every legacy lesson" (e.g. the profile
    screen's lesson-history section) now that content is DB-resident rather
    than discovered from files on disk — the old client never needed this as
    an endpoint because it did that discovery itself via Vite's import.meta.glob."""
    ids = await _legacy_lesson_order(db)
    lessons = []
    for lesson_id in ids:
        content = await db.get(LessonContent, lesson_id)
        material_text = content.materialText if content else None
        parsed = parse_material(material_text or "", [])
        title = _title_from_blocks(parsed["blocks"]) or lesson_label(lesson_id)
        vocab_count = (
            await db.execute(select(func.count()).select_from(VocabularyItem).where(VocabularyItem.lessonId == lesson_id))
        ).scalar_one()
        lessons.append({"lessonId": lesson_id, "title": title, "vocabularyCount": vocab_count})
    return lessons


async def get_lesson_content(db: AsyncSession, lesson_id: str) -> dict:
    content = await db.get(LessonContent, lesson_id)
    vocab_result = await db.execute(
        select(VocabularyItem).where(VocabularyItem.lessonId == lesson_id).order_by(VocabularyItem.position)
    )
    vocabulary = vocab_result.scalars().all()
    q_result = await db.execute(
        select(LessonQuestion).where(LessonQuestion.lessonId == lesson_id).order_by(LessonQuestion.setName, LessonQuestion.position)
    )
    questions = q_result.scalars().all()
    b_result = await db.execute(
        select(LessonBlock).where(LessonBlock.lessonId == lesson_id).order_by(LessonBlock.stage, LessonBlock.position)
    )
    blocks = b_result.scalars().all()

    vocabulary_dtos = [
        {"id": v.id, "german": v.german, "translation": v.translation, "pronunciation": v.pronunciation, "audioUrl": v.audioUrl}
        for v in vocabulary
    ]
    material_text = content.materialText if content else None
    parsed_material = parse_material(material_text or "", vocabulary_dtos)
    taught_before = await _words_taught_before_legacy_lesson(db, lesson_id)
    new_vocabulary = filter_new_vocabulary(vocabulary_dtos, taught_before)

    return {
        "lessonId": lesson_id,
        "materialText": material_text,
        "material": parsed_material["blocks"],
        "phrases": parsed_material["phrases"],
        "videoUrl": content.videoUrl if content else None,
        "audioUrl": content.audioUrl if content else None,
        "vocabulary": vocabulary_dtos,
        "newVocabulary": new_vocabulary,
        "questions": [
            {"setName": q.setName, "prompt": q.prompt, "options": q.options, "correctAnswer": q.correctAnswer}
            for q in questions
        ],
        "blocks": [
            {
                "id": b.id,
                "stage": b.stage,
                "title": b.title,
                "position": b.position,
                "questions": [
                    to_question_dto(q.kind, q.prompt, q.options, q.correctAnswer, q.data)
                    for q in questions
                    if q.blockId == b.id
                ],
            }
            for b in blocks
        ],
        "hasOverrides": content is not None or len(vocabulary) > 0 or len(questions) > 0,
        "updatedAt": to_iso_z(content.updatedAt) if content else None,
    }


async def save_lesson_content(db: AsyncSession, lesson_id: str, payload: ContentPayload, editor_id: str, set_fields: set[str]) -> dict:
    """Saves whichever sections were supplied (set_fields = the request
    body's actually-present top-level keys, since Pydantic can't natively
    distinguish "omitted" from "explicitly null" the way the Zod schema's
    .optional() did). Each section is replaced wholesale in one transaction,
    so a half-applied vocabulary list can never reach learners."""

    if "materialText" in set_fields:
        content = await db.get(LessonContent, lesson_id)
        if content is None:
            db.add(LessonContent(lessonId=lesson_id, materialText=payload.materialText, updatedById=editor_id, updatedAt=utcnow()))
        else:
            content.materialText = payload.materialText
            content.updatedById = editor_id

    if "vocabulary" in set_fields:
        vocabulary = payload.vocabulary or []
        # A word may only belong to one lesson of the course. Reject the
        # whole save (rather than silently dropping a word) and name the
        # lesson the word already lives in, so the admin knows where to look.
        keys = [normalize_word(item.german) for item in vocabulary]
        seen: dict[str, int] = {}
        duplicate_in_payload = None
        for i, key in enumerate(keys):
            if key in seen:
                duplicate_in_payload = key
                break
            seen[key] = i
        if duplicate_in_payload:
            word = vocabulary[keys.index(duplicate_in_payload)].german
            raise DuplicateWordError(f"Слово «{word}» указано в этом уроке дважды")

        if keys:
            clash_result = await db.execute(
                select(VocabularyItem.german, VocabularyItem.lessonId).where(
                    and_(
                        VocabularyItem.courseId == LEGACY_COURSE_ID,
                        VocabularyItem.germanKey.in_(keys),
                        VocabularyItem.lessonId != lesson_id,
                    )
                )
            )
            clashes = clash_result.all()
            if clashes:
                clash_list = "; ".join(f"«{german}» — уже в уроке «{lesson_label(lid)}»" for german, lid in clashes)
                raise DuplicateWordError(f"Это слово уже используется в другом уроке: {clash_list}")

        # Recorded pronunciations are attached to the word, so they survive a
        # full rewrite of the lesson's list.
        existing_result = await db.execute(
            select(VocabularyItem.germanKey, VocabularyItem.audioUrl).where(VocabularyItem.lessonId == lesson_id)
        )
        audio_by_key = dict(existing_result.all())

        old_rows = (await db.execute(select(VocabularyItem).where(VocabularyItem.lessonId == lesson_id))).scalars().all()
        for row in old_rows:
            await db.delete(row)
        await db.flush()

        for index, item in enumerate(vocabulary):
            german_key = normalize_word(item.german)
            pronunciation = item.pronunciation.strip() if item.pronunciation and item.pronunciation.strip() else None
            audio_url = item.audioUrl.strip() if item.audioUrl and item.audioUrl.strip() else audio_by_key.get(german_key)
            db.add(
                VocabularyItem(
                    lessonId=lesson_id,
                    german=item.german,
                    translation=item.translation,
                    pronunciation=pronunciation,
                    audioUrl=audio_url,
                    position=index,
                    germanKey=german_key,
                    courseId=LEGACY_COURSE_ID,
                )
            )

    if "questions" in set_fields:
        questions = payload.questions or []
        old_rows = (await db.execute(select(LessonQuestion).where(LessonQuestion.lessonId == lesson_id))).scalars().all()
        for row in old_rows:
            await db.delete(row)
        await db.flush()

        # Positions are per-set so each stage keeps its own ordering.
        counters: dict[str, int] = {}
        for q in questions:
            position = counters.get(q.setName, 0)
            counters[q.setName] = position + 1
            db.add(
                LessonQuestion(
                    lessonId=lesson_id,
                    setName=q.setName,
                    prompt=q.prompt,
                    options=q.options,
                    correctAnswer=q.correctAnswer,
                    position=position,
                    courseId=LEGACY_COURSE_ID,
                )
            )

    await db.commit()
    return await get_lesson_content(db, lesson_id)


async def set_legacy_lesson_media(db: AsyncSession, lesson_id: str, kind: str, url: str | None) -> dict:
    """Overrides (or clears, with url=None) the bundled video/audio file for
    a legacy lesson. Upserts because a lesson may not have a LessonContent
    row yet."""
    field = "videoUrl" if kind == "video" else "audioUrl"
    content = await db.get(LessonContent, lesson_id)
    if content is None:
        db.add(LessonContent(lessonId=lesson_id, **{field: url}, updatedAt=utcnow()))
    else:
        setattr(content, field, url)
    await db.commit()
    return await get_lesson_content(db, lesson_id)
